### PR TITLE
Save experiment data locally in mobile-app

### PR DESCRIPTION
**Pull request by pjanik**
1. Users can start a new experiment, provide some data, use "back" button and see their saved experiment runs.
2. The app can be reloaded multiple times and the user can load his previously saved experiment run.
3. The second commit moves `useExperiments` to the main app component. It's a bit opinionated, so you can treat it like a proposal. Personally, I like to keep logic and data together in one, top-level component as long as possible. The app component can be treated as a poor replacement for MobX (but for now it's perfectly enough) and other components can have less state and logic. If we ever decide to use some state library, it should be easier to do it as the state won't be spread out in multiple components. Ofc there are some limits. Hooks are useful and sometimes it makes sense to add local state and hooks lower down in the component hierarchy.

### CODE REVIEW

File: src/shared/utils/local-data-storage.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  public load(uuid = ""): DataType | undefined {
+    let data: DataType | undefined;
+    const stringifiedData = window.localStorage.getItem(this.localStorageKeyPrefix + uuid);
+    if (stringifiedData) {
+      try {
+        data = JSON.parse(stringifiedData);
+        // tslint:disable-next-line:no-empty
+      } catch (e) {
```
Comment (kswenson) [inline src/shared/utils/local-data-storage.ts]: If you add a comment in the catch clause, e.g.
```
} catch (e) {
  // ignore errors
}
```
I don't think you'd need the `// tslint:disable-next-line:no-empty`

File: src/mobile-app/components/app.tsx [hunk 2]
```diff
@@ -1,21 +1,90 @@
...
 import css from "./app.module.scss";
 
-export const AppComponent: React.FC<{}> = () => {
+const runsStorage = new LocalDataStorage<IRun>("runs-");
+const availableRunKeysStorage = new LocalDataStorage<string[]>("available-runs");
 
-  const [experiment, setExperiment] = useState<IExperiment|undefined>();
+export const AppComponent: React.FC = () => {
+  const { experiments, upgradeApp } = useExperiments();
+  const [runKey, setRunKey] = useState<string | undefined>();
```
Comment (dougmartin) [inline src/mobile-app/components/app.tsx]: I think that all this code around saving/loading runs should be isolated into a single hook so that it can be tested outside the component and so the component can be focused on rendering the data.

File: src/mobile-app/hooks/use-experiments.test.ts [hunk 3]
```diff
@@ -47,7 +47,8 @@ describe("use-experiments hook", () => {
     const fetchMock = jest.fn().mockRejectedValue(new Error("test fetch failure"));
     (window as any).fetch = fetchMock;
 
-    const mockedStorage: IExperimentStorage = {
+    const mockedStorage = {
```
Comment (dougmartin) [inline src/mobile-app/hooks/use-experiments.test.ts]: Why did you remove the typing?  We lose compile time checks in the tests doing that.

Comment (dougmartin) [reply src/mobile-app/hooks/use-experiments.test.ts]: Ah, I see you changed to the LocalDataStorage.  I think you should create an interface that that class implements and then export it so that we can keep the type checking here.

File: src/mobile-app/components/app.tsx [hunk 4]
```diff
@@ -1,21 +1,90 @@
...
 
 import css from "./app.module.scss";
 
-export const AppComponent: React.FC<{}> = () => {
+const runsStorage = new LocalDataStorage<IRun>("runs-");
+const availableRunKeysStorage = new LocalDataStorage<string[]>("available-runs");
 
-  const [experiment, setExperiment] = useState<IExperiment|undefined>();
+export const AppComponent: React.FC = () => {
+  const { experiments, upgradeApp } = useExperiments();
```
Comment (dougmartin) [inline src/mobile-app/components/app.tsx]: I don't mind that the experiments array was "lifted" out of the component.  I did put it there for a couple of reasons:

1) to keep the data at the closest component that needed it (here you use it but then only pass it to the single component - it isn't shared by any other ones) 

2) so that the json was reloaded each time the picker was shown (thinking again about this it doesn't make operational sense as the students won't be switching between experiments very often to trigger this reload).

File: src/mobile-app/components/app.tsx [hunk 5]
```diff
@@ -1,21 +1,90 @@
...
+import { LocalDataStorage } from "../../shared/utils/local-data-storage";
+import { RunPicker } from "./run-picker";
+import { IRun } from "./types";
+import { useExperiments } from "../hooks/use-experiments";
 
 import css from "./app.module.scss";
 
-export const AppComponent: React.FC<{}> = () => {
+const runsStorage = new LocalDataStorage<IRun>("runs-");
+const availableRunKeysStorage = new LocalDataStorage<string[]>("available-runs");
```
[the remark this answers is not in the document]
Comment (pjanik) [reply src/mobile-app/components/app.tsx]: The argument is only the local storage key (or prefix if you use IDs), so that's always a string. It's not the initial value.

File: src/mobile-app/components/app.tsx (hunk 2 above)
Comment (pjanik) [reply src/mobile-app/components/app.tsx]: Right, I thought about it after I wrote that code. I'll try to move it to a hook.

File: src/mobile-app/hooks/use-experiments.test.ts (hunk 3 above)
Comment (pjanik) [reply src/mobile-app/hooks/use-experiments.test.ts]: Right, because now it's a different class and this interface is gone. We get type checking when this object is passed to `useExperiments` a few lines below (as this hook is expecting specific type as an argument). That's why I was forced to add `localStorageKeyPrefix`. 
Adding a type here would be only helpful if this mock is never used I think. So, in practice not very useful. But I can add `const mockedStorage: LocalDataStorage<Experiments> =`. Or add a new interface that this storage implements that e.g. doesn't require `localStorageKeyPrefix`.

File: src/mobile-app/components/app.tsx (hunk 4 above)
Comment (pjanik) [reply src/mobile-app/components/app.tsx]: I get it. If we didn't have hooks, I'd definitely prefer to keep as much of the logic as we can in the top-level component, as then it's in one place. When I lifted it out of this component, I did it for that reason. But I did it kinda forgetting that hooks keep the logic out of the component already. They add some state quite often, but most of the complexity is under the hood. Also, the logic is easier to find as they're kept together etc. On the other hand, the whole flow of the app and various state handling will be spread across many components. But we can avoid unnecessary props passing (which is annoying long term) and it's more likely we can avoid the state handling library in the end. So, I have no strong opinion here. Probably we should just pick one approach and stick to it. I think your approach is fine because of the hook advantages. I'll move the hook back here.

**Comment by pjanik** [comment]
Thanks, agree on hooks refactoring, I'll do it.

**Comment by pjanik** [comment]
I think all the comments are addressed in PR https://github.com/concord-consortium/vortex/pull/14 (a new version of this one).
